File: app/src/main/java/com/br/example/fakebank/presentations/viewModels/StatementViewModel.java
package com.br.example.fakebank.presentations.viewModels;

import androidx.lifecycle.ViewModel;

import com.br.example.fakebank.domains.services.StatementService;
import com.br.example.fakebank.presentations.Erros.ErrorUtils;
import com.br.example.fakebank.presentations.Erros.InvalidApiAccessError;
import com.br.example.fakebank.presentations.handles.StatementHandle;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

public class StatementViewModel extends ViewModel {

    private StatementService currencyService;
    private StatementHandle statementHandle;

    private CompositeDisposable disposable = new CompositeDisposable();

    public StatementViewModel(StatementService currencyService, StatementHandle statementHandle) {
        this.currencyService = currencyService;
        this.statementHandle = statementHandle;
    }

    public void listStatements(Integer userId) {
        statementHandle.setLoading(true);
        disposable.add(currencyService.listStatements(userId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        greeting -> {
                            if (greeting.getStatementList() != null){
                                statementHandle.reloadListStatement(greeting.getStatementList());
                            }else{
                                statementHandle.showError(new ErrorUtils(greeting.getError().getMessage()));
                            }
                            statementHandle.setLoading(false);
                        },
                        throwable -> {
                            statementHandle.showError(new InvalidApiAccessError());
                            statementHandle.setLoading(false);
                        }
                )
        );
    }

    public void didClickLogout(){
        statementHandle.actionLogout();
    }
}

File: SantanderTestValber/app/src/main/java/com/valber/santandertestvalber/presentation/StatementViewModel.kt
package com.valber.santandertestvalber.presentation

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.valber.data.model.Statement
import com.valber.domain2.StatementCase
import com.valber.santandertestvalber.extensions.setError
import com.valber.santandertestvalber.extensions.setLoading
import com.valber.santandertestvalber.extensions.setSuccess
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers

class StatementViewModel(private val statementCase: StatementCase) : ViewModel() {

    val statement = MutableLiveData<Resource<List<Statement>>>()
    private val compositeDisposable = CompositeDisposable()

    fun get(id: Int) =
        compositeDisposable.add(
            statementCase.getStatements(id)
                .doOnSubscribe { statement.setLoading() }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                    { statement.setSuccess(it) },
                    { statement.setError(it.message) }
                )
        )

    override fun onCleared() {
        compositeDisposable.dispose()
        super.onCleared()
    }
}